### PR TITLE
[Lib] Add support for creating private registry

### PR DESCRIPTION
**Pull request by pranavprakash20**
Adding supporting lib for performing disconnected installation

# Description

Please include Automation development guidelines. Source of Test case - New Feature/Regression Test/Close loop of customer BZs
<details>

<summary>click to expand checklist</summary>

- [ ] Create a test case in Polarion reviewed and approved.
- [ ] Create a design/automation approach doc. Optional for tests with similar tests already automated.
- [ ] Review the automation design
- [ ] Implement the test script and perform test runs
- [ ] Submit PR for code review and approve
- [ ] Update Polarion Test with Automation script details and update automation fields
- [ ] If automation is part of Close loop, update BZ flag qe-test_coverage “+” and link Polarion test
</details>

### CODE REVIEW

File: cephci/utils/configure.py [hunk 1]
```diff
@@ -76,3 +80,180 @@ def setup_ssh_keys(installer, nodes):
...
+
+def accept_eula(node):
+    cmd = "cat" if accept_eula else "touch"
+    cmd += " /usr/share/ibm-storage-ceph-license/accept"
+
+    Package(node).install("ibm-storage-ceph-license", env_vars=accept_eula)
+    return node.exec_command(sudo=True, cmd=cmd, check_ec=True)
+
+
+def exec_cephadm_preflight(installer, build_type, ceph_repo):
```
Comment (vamahaja) [inline cephci/utils/configure.py:237]: As preflight playbook is not part of disconnected install better to send separate PR

File: cephci/utils/configure.py [hunk 2]
```diff
@@ -76,3 +80,180 @@ def setup_ssh_keys(installer, nodes):
...
+    """Copy the certificate to anotheer node in the cluster"""
+    cmd = f"scp /opt/registry/certs/domain.crt root@{secondary_node[1].hostname}:/etc/pki/ca-trust/source/anchors/"
+    out, _ = node.exec_command(cmd=cmd, sudo=True)
+    if out:
+        log.error("Failed to copy the certificate to another node")
+        return False
+    return True
+
+
+def start_local_private_registry(node):
```
Comment (vamahaja) [inline cephci/utils/configure.py:185]: As discussed add new class 'Container' in `cli/utilities/containers.py`  with run method.

File: cephci/utils/configure.py [hunk 3]
```diff
@@ -76,3 +80,180 @@ def setup_ssh_keys(installer, nodes):
...
+        log.error("Failed to update-ca-trust")
+        return False
+
+    # Update the ca-trust
+    if not update_ca_trust(node):
+        return False
+    return True
+
+
+def update_ca_trust(node):
```
Comment (vamahaja) [inline cephci/utils/configure.py:155]: `update_ca_trust` & `validate_trusted_list` can be merged with method name `setup_ca_trust`

File: cephci/utils/configure.py [hunk 4]
```diff
@@ -76,3 +80,180 @@ def setup_ssh_keys(installer, nodes):
...
             ),
         )
+
+
+def create_private_registry_dirs(node):
+    """
+    Create private registry directories for disconnected install
+    """
+    pass
+    cmd = "mkdir -p /opt/registry/{auth,certs,data}"
```
[the remark this answers is not in the document]
Comment (pranavprakash20) [reply cephci/utils/configure.py:90]: Noted

File: cephci/utils/configure.py (hunk 3 above)
Comment (pranavprakash20) [reply cephci/utils/configure.py:154]: The above methods just performs a validation of certificate creation. So clubbing them to `setup_ca_trust` will give a wrong interpretation of what these methods do.

File: cephci/utils/configure.py (hunk 2 above)
Comment (pranavprakash20) [reply cephci/utils/configure.py:184]: Done

File: cephci/prereq.py [hunk 5]
```diff
@@ -44,6 +61,8 @@
         -k --setup-ssh-keys <BOOL>  Setup SSH keys on cluster
         -f --config <FILE>          CephCI configuration file
         -l --log-level <LOG>        Log level for log utility
+        -p --private-registry <BOOL> Use private registry
+        -cp --cephadm-preflight <BOOL> Run cephadm preflight
```
[the remark this answers is not in the document]
Comment (pranavprakash20) [reply cephci/prereq.py:65]: Done

File: ceph/ceph.py [hunk 6]
```diff
@@ -2107,6 +2107,15 @@ def download_file(self, src, dst, sudo=False):
         client = self.rssh if sudo else self.ssh
         client().open_sftp().get(src, dst)
 
+    def create_dirs(self, dir_path, sudo=False):
+        """Create directory on node
+        Args:
+            dir_path (str): Directory path to create
+            sudo (bool): Use root access
+        """
+        cmd = "mkdir -p {}".format(dir_path)
```
Comment (vamahaja) [inline ceph/ceph.py:2116]: can we use [1] instead 
[1] https://docs.paramiko.org/en/stable/api/sftp.html#paramiko.sftp_client.SFTPClient.mkdir

File: cephci/prereq.py [hunk 7]
```diff
@@ -142,9 +163,84 @@ def enable_rhel_repos(node, server, distro):
     return True
 
 
-def prereq(cluster, build, subscription, registry, ssh=False):
+def setup_private_container_registry(installer, nodes, registry, build_type):
+    """
+    Performs the pre-reqs required for the disconnected install
+    """
+    reg_uname = "Test"
```
Comment (vamahaja) [inline cephci/prereq.py:170]: Create global const variables for these creds

File: cli/utilities/containers.py [hunk 8]
```diff
@@ -41,3 +41,21 @@ def login(
...
+
+
+class Container(Cli):
+    """This module provides CLI interface for performing ops in containers"""
+
+    def __init__(self, nodes, package="podman"):
+        super(Container, self).__init__(nodes)
+        self.base_cmd = package
+
+    def run(self, command):
```
Comment (vamahaja) [inline cli/utilities/containers.py:53]: This is not in cli format.

**Review by vamahaja — CHANGES_REQUESTED** [review]

File: cli/utilities/containers.py (hunk 8 above)
Comment (pranavprakash20) [reply cli/utilities/containers.py:45]: updated

**Comment by pranavprakash20** [comment]
commets addressed. @vamahaja

File: cephci/prereq.py [hunk 9]
```diff
@@ -22,6 +35,9 @@
 
 log = Log(__name__)
 
+reg_uname = "Test"
```
Comment (vamahaja) [inline cephci/prereq.py:38]: Golabal variables should be in capital letters `REGISTRY_USERNAME`

File: cli/utilities/containers.py [hunk 10]
```diff
@@ -41,3 +45,81 @@ def login(
...
+            env (list): List of environment variables
+            volume (list): List of volumes
+            ports (list): List of ports
+            rm (list): List of entries to be removed
+            detach (bool): Detach flag
+            restart (str): Restart flag
+            cmds (str): Other commands to be executed
+        """
+        if not image and not rm:
+            raise ContainerError("Image name or rm option needs to be provided")
```
Comment (vamahaja) [inline cli/utilities/containers.py:81]: instead use this exception - https://github.com/red-hat-storage/cephci/blob/master/cli/exceptions.py#L7

File: cephci/utils/configure.py [hunk 11]
```diff
@@ -76,3 +80,190 @@ def setup_ssh_keys(installer, nodes):
...
+        restart="always",
+        detach_key=detach,
+    ):
+        log.error("Failed to start local secure private registry")
+        return False
+    return True
+
+
+def add_images_to_private_registry(
+    node, registry_uname, registry_passwd, registry, build_type
```
Comment (vamahaja) [inline cephci/utils/configure.py:218]: use variables `username` and `password`

File: cephci/utils/configure.py [hunk 12]
```diff
@@ -76,3 +80,190 @@ def setup_ssh_keys(installer, nodes):
...
+    out, _ = node.exec_command(cmd=cmd, sudo=True)
+    if node.hostname not in out:
+        log.error("Trust list not updated")
+        return False
+    return True
+
+
+def copy_cert_to_secondary_node(node, secondary_node):
+    """Copy the certificate to anotheer node in the cluster"""
+    cmd = f"scp /opt/registry/certs/domain.crt root@{secondary_node[1].hostname}:/etc/pki/ca-trust/source/anchors/"
```
Comment (vamahaja) [inline cephci/utils/configure.py:176]: Create global const variables for path

File: cephci/utils/configure.py [hunk 13]
```diff
@@ -76,3 +80,190 @@ def setup_ssh_keys(installer, nodes):
...
+    out, _ = node.exec_command(cmd=cmd, sudo=True)
+    if out:
+        log.error("Failed to create a link to the domain certificate")
+        return False
+    return True
+
+
+def add_cert_to_trusted_list(node):
+    """Add the certificate to the trusted list on the private registry node"""
+    cmd = "cp /opt/registry/certs/domain.crt /etc/pki/ca-trust/source/anchors/"
```
Comment (vamahaja) [inline cephci/utils/configure.py:142]: Create global const variables for path

File: cephci/utils/configure.py [hunk 14]
```diff
@@ -76,3 +80,190 @@ def setup_ssh_keys(installer, nodes):
...
+    out, _ = node.exec_command(cmd=cmd, sudo=True)
+    if out:
+        log.error("Failed to create a self-signed certificate")
+        return False
+    return True
+
+
+def create_link_to_domain_cert(node):
+    """Creates a link to the domain certificate"""
+    cmd = "ln -s /opt/registry/certs/domain.crt /opt/registry/certs/registry.crt"
```
Comment (vamahaja) [inline cephci/utils/configure.py:132]: Create global const variables for path

File: cephci/utils/configure.py [hunk 15]
```diff
@@ -76,3 +80,190 @@ def setup_ssh_keys(installer, nodes):
...
+    if out:
+        log.error("Failed to create credentials for accessing the private registry")
+        return False
+    return True
+
+
+def create_self_signed_certificate(node):
+    """Creates self-signed certificate for private registry"""
+    cmd = (
+        f"openssl req -newkey rsa:4096 -nodes -sha256 -keyout /opt/registry/certs/domain.key -x509 -days 365 -out "
```
Comment (vamahaja) [inline cephci/utils/configure.py:119]: Create global const variables for path

File: cephci/utils/configure.py [hunk 16]
```diff
@@ -76,3 +80,190 @@ def setup_ssh_keys(installer, nodes):
...
+    Create credentials for accessing the private registry
+    Args:
+        node (ceph): Node to execute cmd
+        username (str): Username for the private registry
+        password (str): Password for the private registry
+
+    Returns:
+
+    """
+    cmd = f"htpasswd -bBc /opt/registry/auth/htpasswd {username} {password}"
```
Comment (vamahaja) [inline cephci/utils/configure.py:108]: Create global const variables for path

**Review by vamahaja — CHANGES_REQUESTED** [review]

**Comment by pranavprakash20** [comment]
All comments addressed @vamahaja

File: cephci/prereq.py [hunk 17]
```diff
@@ -44,6 +63,10 @@
         -k --setup-ssh-keys <BOOL>  Setup SSH keys on cluster
         -f --config <FILE>          CephCI configuration file
         -l --log-level <LOG>        Log level for log utility
+        -p --private-registry <BOOL> Use private registry
```
Comment (vamahaja) [inline cephci/prereq.py:66]: This will change indentation for existing parameters.

File: cephci/utils/configure.py [hunk 18]
```diff
@@ -76,3 +85,188 @@ def setup_ssh_keys(installer, nodes):
...
+        log.error("Failed to start local secure private registry")
+        return False
+    return True
+
+
+def add_images_to_private_registry(node, username, password, registry, build_type):
+    """Red Hat Ceph Storage 5 image, Prometheus images, and Dashboard image from the Red Hat Customer Portal
+    to the private registry:"""
+
+    rm = "registry.redhat.io/rhel8/skopeo:8.5-8"
```
Comment (vamahaja) [inline cephci/utils/configure.py:226]: Add golbal constant for `skepeo` image

File: cephci/utils/configure.py [hunk 19]
```diff
@@ -76,3 +85,188 @@ def setup_ssh_keys(installer, nodes):
...
+    return True
+
+
+def start_local_private_registry(node):
+    """Start local private registry"""
+
+    image = "myprivateregistry"
+    ports = "5000:5000"
+    volumes = [
+        "/opt/registry/data:/var/lib/registry:z",
```
Comment (vamahaja) [inline cephci/utils/configure.py:195]: Better to create golabl constants for these paths

File: cephci/utils/configure.py [hunk 20]
```diff
@@ -76,3 +85,188 @@ def setup_ssh_keys(installer, nodes):
...
+    return True
+
+
+def add_images_to_private_registry(node, username, password, registry, build_type):
+    """Red Hat Ceph Storage 5 image, Prometheus images, and Dashboard image from the Red Hat Customer Portal
+    to the private registry:"""
+
+    rm = "registry.redhat.io/rhel8/skopeo:8.5-8"
+    volumes = [
+        "/opt/registry/certs:/certs:Z",
```
Comment (vamahaja) [inline cephci/utils/configure.py:228]: Same as above. It will be reused here

File: cli/utilities/containers.py [hunk 21]
```diff
@@ -33,11 +30,89 @@ def login(
         elif username and password:
             cmd += f" --username {username} --password {password}"
         else:
-            raise ContainerRegistryError("Authentication details needs to be provided")
```
Comment (vamahaja) [inline cli/utilities/containers.py:36]: Don't include this here, will have separate PR for this change

File: cephci/utils/configure.py (hunk 3 above)
Comment (vamahaja) [reply cephci/utils/configure.py:159]: +1

File: cli/utilities/containers.py (hunk 21 above)
Comment (pranavprakash20) [reply cli/utilities/containers.py:36]: Moved out of this PR

File: cephci/utils/configure.py (hunk 20 above)
Comment (pranavprakash20) [reply cephci/utils/configure.py]: Noted, changed

File: cephci/utils/configure.py (hunk 19 above)
Comment (pranavprakash20) [reply cephci/utils/configure.py]: Created global variable for all the paths

File: cephci/utils/configure.py (hunk 18 above)
Comment (pranavprakash20) [reply cephci/utils/configure.py]: Skopeo global variable added

File: cephci/prereq.py (hunk 17 above)
Comment (pranavprakash20) [reply cephci/prereq.py]: Indentation fixed for all

File: cli/utilities/containers.py (hunk 21 above)
Comment (vamahaja) [reply cli/utilities/containers.py:36]: ????
Still changes are here

File: cephci/utils/configure.py [hunk 22]
```diff
@@ -19,6 +22,18 @@
 
 CHMOD_CONFIG = f"chmod 600 {SSH_CONFIG}"
 SSHPASS = "sshpass -p {}"
+REGISTRY_AUTH_PATH = "/opt/registry/auth/htpasswd"
```
Comment (vamahaja) [inline cephci/utils/configure.py:25]: Here `/opt/registry` is common path for most of the time. Use below format -

- Create variable for common format as `REGISTRY_PATH = "/opt/registry"`
- Use `REGISTRY_PATH` to create other variables like - `DOMAIN_KEY_PATH = f{REGISTRY_PATH}/certs/domain.key`

**Review by vamahaja — CHANGES_REQUESTED** [review]

**Comment by vamahaja** [comment]
Share results

**Review by vamahaja — CHANGES_REQUESTED** [review]

File: cephci/prereq.py [hunk 23]
```diff
@@ -184,18 +263,24 @@ def prereq(
...
+        # cephadm preflight is required to have ssh keys setup
         setup_ssh_keys(installer, nodes)
 
     if cephadm_ansible and not Rpm(installer).query(CEPHADM_ANSIBLE):
         Package(installer).install(CEPHADM_ANSIBLE, nogpgcheck=True)
+
     if cephdam_preflight:
         if not (ssh or cephadm_ansible):
-            raise ConfigError(
+            raise NodeConfigError(
```
Comment (vamahaja) [inline cephci/prereq.py:275]: `NodeConfigError` ???

File: cephci/prereq.py [hunk 24]
```diff
@@ -184,18 +263,24 @@ def prereq(
         if registry != "skip":
             registry_login(node, registry, build)
 
-    if ssh:
+    if ssh or cephdam_preflight:
+        # cephadm preflight is required to have ssh keys setup
         setup_ssh_keys(installer, nodes)
 
     if cephadm_ansible and not Rpm(installer).query(CEPHADM_ANSIBLE):
```
Comment (vamahaja) [inline cephci/prereq.py:270]: This need to be changed now -
```
if (cephadm_ansible or cephdam_preflight) and not Rpm(installer).query(CEPHADM_ANSIBLE):
```

**Review by vamahaja — CHANGES_REQUESTED** [review]

File: cephci/prereq.py [hunk 25]
```diff
@@ -150,19 +169,79 @@ def enable_rhel_repos(node, server, distro):
     return True
 
 
+def setup_private_container_registry(installer, nodes, registry, build_type):
+    """
```
Comment (adityaramteke) [inline cephci/prereq.py:173]: Pranav, can we please describe the method argument?

File: cephci/prereq.py [hunk 26]
```diff
@@ -150,19 +169,79 @@ def enable_rhel_repos(node, server, distro):
...
+    if not add_cert_to_trusted_list(installer):
+        return False
+
+    # Validate the truster list is updated
+    if not validate_trusted_list(installer):
+        return False
+
+    # Step 6: Copy the certificate to any nodes that will access the private registry for installation and update the
+    # trusted list
+    if not copy_cert_to_secondary_node(installer, nodes[1]):
```
Comment (adityaramteke) [inline cephci/prereq.py:204]: We'll need to copy the cert to all the nodes in the cluster!

**Review by adityaramteke — CHANGES_REQUESTED** [review]
